### PR TITLE
workaround for loading in images and image validation

### PR DESCRIPTION
**Pull request by vhart**
@mesbekmek i've added in some workarounds here and it loads in normally for the most part just a bug now where it doesn't immediately resize, maybe you can take a stab at that?
It loads in normally, then when you scroll up it resizes to the actually desired size - havent ironed out why just yet, but will look into it when I get a few minutes (going to the florist lol)

### CODE REVIEW

**Comment by mesbekmek** [comment]
@vhart Pulled your changes locally, and the image resizing looks good, although during scroll/ fast scroll it takes the images a much longer time to load. I also see the initial wrong image size that you mentioned.

What would you say is the advantage of using the collection view compared to the table view?

File: Meme/Meme/ImageCache.swift [hunk 1]
```diff
@@ -16,12 +16,17 @@ protocol ImageCaching {
...
     static let shared = ImageCache()
     private var imageDictionary: [URL:UIImage] = [:]
+    private let lock = NSLock()
 
     func getImage(for url: URL) -> UIImage? {
         return imageDictionary[url]
     }
 
     func setImage(image: UIImage, for url: URL) {
+        lock.lock()
```
Comment (mesbekmek) [inline Meme/Meme/ImageCache.swift]: Is the idea with the lock that you don't want any `setImage` calls to happen while one is already in progress? Also does the lock affect the getter or only the setter?

File: Meme/Meme/MemeTemplatesViewController.swift [hunk 2]
```diff
@@ -121,32 +107,32 @@ class MemeCollectionViewFlowLayout: UICollectionViewFlowLayout {
         fatalError("init(coder:) has not been implemented")
     }
 
-    override func shouldInvalidateLayout(forBoundsChange newBounds: CGRect) -> Bool {
-        return true
-    }
+//    override func shouldInvalidateLayout(forBoundsChange newBounds: CGRect) -> Bool {
```
Comment (mesbekmek) [inline Meme/Meme/MemeTemplatesViewController.swift]: Uncommenting out this function helps in resizing the images correctly while scrolling, you no longer need to scroll up to get the correctly sized image. You merely need to start scrolling. I think it has to do with the fact that it `causes the collection view to requery the layout for geometry information` according to the documentation.

It also helped the image loading in, keep up with the fast scrolling. I didn't have to wait as long to get the correctly sized images to come in, even when I scroll super fast.

File: Meme/Meme/ImageCache.swift (hunk 1 above)
Comment (vhart) [reply Meme/Meme/ImageCache.swift]: The idea is to synchronize the setter since it's being called on different threads. The getter could also be synchronized, but in this case we don't need to.. I think - since the getter is only accessed on the main thread (as it is right now)

File: Meme/Meme/MemeTemplatesViewController.swift (hunk 2 above)
Comment (vhart) [reply Meme/Meme/MemeTemplatesViewController.swift]: yeah this fixed a crash I was running into - good call

**Comment by vhart** [comment]
@mesbekmek 
> What would you say is the advantage of using the collection view compared to the table view?

In my opinion collection views are a super set of table views and offer a lot more customization. In this case I suggested it as an added complexity so we could learn something new. However, it seems that self sizing cells in collection views still aren't _perfected_ and there are some unfortunate breaking changes that require workarounds as you've seen in this branch. That being said I think the act of having to try something like this in a new language offers a good learning opportunity.

The latest commit fixes the wrong size on initial load problem, but I feel like there may be performance concerns.

**Comment by vhart** [comment]
@mesbekmek sorry to derail the initial point of this app 😓 On the bright side I think this branch finally has a working version of the resizing without the choppy loading and scrolling - took many medium articles and resorting to some size caching. Feel free to use it as reference or to scrap it all together and go back to the table view if it's not worth the work!

**Comment by mesbekmek** [comment]
@vhart No worries at all. I think it's a good exercise to learn this for the both of us. However, there are quite a lot of changes, so before merging this in I'll try to post comments on different parts the PR to ensure I understand everything and we're on the same page.

File: Meme/Meme/ImageCache.swift [hunk 3]
```diff
@@ -8,20 +8,210 @@
...
+//        lock.lock()
+//        defer {
+//            lock.unlock()
+//        }
+//        imageDictionary[url] = image
+//    }
+//}
+
+/*
+ See LICENSE folder for this sample’s licensing information.
```
Comment (mesbekmek) [inline Meme/Meme/ImageCache.swift:35]: Was the AsyncFetcher class taken from some external codebase or did you write it yourself? Kinda curious as to why this comment was added?

File: Meme/Meme/ImageCache.swift [hunk 4]
```diff
@@ -8,20 +8,210 @@
...
+    private let serialAccessQueue = OperationQueue()
+
+    /// An `OperationQueue` that contains `AsyncFetcherOperation`s for requested data.
+    private let fetchQueue = OperationQueue()
+
+    /// A dictionary of arrays of closures to call when an object has been fetched for an id.
+    private var completionHandlers = [URL: [(UIImage?) -> Void]]()
+
+    /// An `NSCache` used to store fetched objects.
+    private var cache = NSCache<NSURL, UIImage>()
```
Comment (mesbekmek) [inline Meme/Meme/ImageCache.swift:57]: Did you decide to use an NSCache instance because it takes care of locking for you automatically? I'm referring to this from the Apple documentation:

`You can add, remove, and query items in the cache from different threads without having to lock the cache yourself.`

Or was there another benefit to using NSCache as well?

Comment (mesbekmek) [reply Meme/Meme/ImageCache.swift:57]: Also I think it can be a let property as you're not mutating it after declaring it, just accessing its public getters and setters.

File: Meme/Meme/ImageCache.swift [hunk 5]
```diff
@@ -8,20 +8,210 @@
...
+
+    static let shared = AsyncFetcher()
```
Comment (mesbekmek) [inline Meme/Meme/ImageCache.swift:59]: This is how singleton's work in Swift right? No, dispatch_once token business necessary?

File: Meme/Meme/ImageCache.swift [hunk 6]
```diff
@@ -8,20 +8,210 @@
...
+
+    // MARK: Initialization
+
+    init() {
+        serialAccessQueue.maxConcurrentOperationCount = 1
```
Comment (mesbekmek) [inline Meme/Meme/ImageCache.swift:64]: This is to ensure that there is only one operation executing its task in the serial queue at any given them, right?

File: Meme/Meme/ImageCache.swift (hunk 3 above)
Comment (vhart) [reply Meme/Meme/ImageCache.swift:35]: Good question! Apple had a sample project where they demonstrated prefetching and this was in it, sorta. It needed some small changes but what was nice was that it had the operation queues already set up for cancellation.

File: Meme/Meme/ImageCache.swift (hunk 4 above)
Comment (vhart) [reply Meme/Meme/ImageCache.swift:57]: Good call and also yeah that was the main reason! I had it in my branch and when I switched to the async fetcher they had one also so I just changed the type.

File: Meme/Meme/ImageCache.swift (hunk 5 above)
Comment (vhart) [reply Meme/Meme/ImageCache.swift:59]: Correct. Static enforces that there’s just one

File: Meme/Meme/ImageCache.swift [hunk 7]
```diff
@@ -8,20 +8,210 @@
...
+     - completion: An optional called when the data has been fetched.
+     */
+    func fetchAsync(_ url: URL, completion: ((UIImage?) -> Void)? = nil) {
+        // Use the serial queue while we access the fetch queue and completion handlers.
+        serialAccessQueue.addOperation {
+            // If a completion block has been provided, store it.
+            if let completion = completion {
+                let handlers = self.completionHandlers[url, default: []]
+                self.completionHandlers[url] = handlers + [completion]
+            }
```
Comment (mesbekmek) [inline Meme/Meme/ImageCache.swift:83]: The syntax in the two lines above is a bit confusing to me. Here's what I understand below, and please correct me if I'm wrong:

1. On line 81, we set a variable called handlers (which should be of type: array of completion handlers) to the value of a given url that was passed into this function. 
My question here is why do you have to do `default:[]` at the end there? Couldn't you just do `self.completionHandlers[url]`?

2. Line 82, seems like it's just appending the array containing the passed in completion block to the array of handlers from the previous line.

File: Meme/Meme/ImageCache.swift (hunk 6 above)
Comment (vhart) [reply Meme/Meme/ImageCache.swift:64]: Correct. The operation queue inherently is concurrent, this allows us to limit it to one operation at a time which essentially gives us a serial queue. It’s equivalent in some ways to using a dispatch barrier on each call, but this is cleaner

File: Meme/Meme/ImageCache.swift [hunk 8]
```diff
@@ -8,20 +8,210 @@
...
+
+            self.fetchData(for: url)
+        }
+    }
+
+    /**
+     Returns the previously fetched data for a specified `URL`.
```
Comment (mesbekmek) [inline Meme/Meme/ImageCache.swift:90]: Why `previously` fetched data? Shouldn't it just be returns the fetched data? Or am I hair-splitting here?

Comment (mesbekmek) [reply Meme/Meme/ImageCache.swift:90]: Nvm, looking at the code makes sense. It's cause it's asking the cache.

File: Meme/Meme/ImageCache.swift [hunk 9]
```diff
@@ -8,20 +8,210 @@
...
+
+    /**
+     Cancels any enqueued asychronous fetches for a specified `URL`. Completion
+     handlers are not called if a fetch is canceled.
+
+     - Parameter url: The `URL` to cancel fetches for.
+     */
+    func cancelFetch(_ url: URL) {
+        serialAccessQueue.addOperation {
+            self.fetchQueue.isSuspended = true
```
Comment (mesbekmek) [inline Meme/Meme/ImageCache.swift:107]: Do we always have to suspend the entire queue before canceling a specific operation inside of it?

File: Meme/Meme/ImageCache.swift (hunk 7 above)
Comment (vhart) [reply Meme/Meme/ImageCache.swift:83]: Dictionary access returns an optional, the default here is syntactic sugar. It’s the equivalent of
completionHandlers[url] ?? []
(Nil coalescing)

And yeah that’s exactly what it it’s doing on 82

Comment (mesbekmek) [reply Meme/Meme/ImageCache.swift:83]: Maybe for question 1 above, the syntax is saying, when trying to access the completion handler for a given url in the dictionary, if it comes up empty, assign my value to an empty array instead?

File: Meme/Meme/MemeTemplatesCollectionViewCell.swift [hunk 10]
```diff
@@ -10,49 +10,138 @@ import UIKit
...
-    lazy var memeImageView: UIImageView = {
-        var imageView = UIImageView()
+    static let sizingCell = MemeTemplatesCollectionViewCell(frame: .zero)
+
+    lazy var memeImageView: AsyncImageView = {
+        var imageView = AsyncImageView()
+        imageView.onAsyncImageSet = { [weak self] in self?.updateLayout() }
         imageView.translatesAutoresizingMaskIntoConstraints = false
         imageView.clipsToBounds = true
+        imageView.layer.masksToBounds = true
```
Comment (mesbekmek) [inline Meme/Meme/MemeTemplatesCollectionViewCell.swift:20]: Was there a specific sub-layer of the image view you were trying to clip when setting this to true? Or did you add it for another reason?

File: Meme/Meme/MemeTemplatesCollectionViewCell.swift [hunk 11]
```diff
@@ -10,49 +10,138 @@ import UIKit
...
+            guard let maxWidth = maxWidth else {
                 return
             }
-            maxHeightConstraint.isActive = true
-            maxHeightConstraint.constant = maxHeight
+            maxWidthConstraint.isActive = true
+            maxWidthConstraint.constant = maxWidth
+            memeImageView.maxWidth = maxWidth
+            self.contentView.setNeedsLayout()
+            self.contentView.layoutIfNeeded()
```
Comment (mesbekmek) [inline Meme/Meme/MemeTemplatesCollectionViewCell.swift:36]: Is calling both one after the other necessary?

File: Meme/Meme/MemeTemplatesCollectionViewCell.swift (hunk 10 above)
Comment (vhart) [reply Meme/Meme/MemeTemplatesCollectionViewCell.swift:20]: This may not be necessary. I think I forgot it in. I added it very early on and it didn’t solve the problem I thought it would

File: Meme/Meme/MemeTemplatesViewController.swift [hunk 12]
```diff
@@ -76,38 +75,74 @@ class MemeTemplatesViewController: UIViewController {
...
+    
+    override func viewWillLayoutSubviews() {
+        super.viewWillLayoutSubviews()
+        collectionView.collectionViewLayout.invalidateLayout()
+    }
+}
+
+extension MemeTemplatesViewController: UICollectionViewDataSourcePrefetching {
+    func collectionView(_ collectionView: UICollectionView, prefetchItemsAt indexPaths: [IndexPath]) {
+        let imageURLs = indexPaths.compactMap { templates[$0.row].imageURL }
```
Comment (mesbekmek) [inline Meme/Meme/MemeTemplatesViewController.swift:89]: So, just to be clear this returns an array of image URLs corresponding to templates at the given indexPaths?

File: Meme/Meme/MemeTemplatesViewController.swift [hunk 13]
```diff
@@ -76,38 +75,74 @@ class MemeTemplatesViewController: UIViewController {
...
+
+    func collectionView(_ collectionView: UICollectionView, cancelPrefetchingForItemsAt indexPaths: [IndexPath]) {
+        let imageURLs = indexPaths.compactMap { templates[$0.row].imageURL }
+        for url in imageURLs { AsyncFetcher.shared.cancelFetch(url) }
+    }
 }
 
 extension MemeTemplatesViewController: UICollectionViewDataSource, UICollectionViewDelegate, UICollectionViewDelegateFlowLayout {
     func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        collectionView.layoutIfNeeded()
```
Comment (mesbekmek) [inline Meme/Meme/MemeTemplatesViewController.swift]: Why is this necessary?

File: Meme/Meme/MemeTemplatesViewController.swift [hunk 14]
```diff
@@ -76,38 +75,74 @@ class MemeTemplatesViewController: UIViewController {
...
         return templates.count
     }
 
     func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        collectionView.layoutIfNeeded()
```
Comment (mesbekmek) [inline Meme/Meme/MemeTemplatesViewController.swift]: Same as above, why is this necessary?

File: Meme/Meme/ImageCache.swift (hunk 9 above)
Comment (vhart) [reply Meme/Meme/ImageCache.swift:107]: Suspending the queue prevents additional operations from being started.

File: Meme/Meme/MemeTemplatesCollectionViewCell.swift [hunk 15]
```diff
@@ -10,49 +10,138 @@ import UIKit
 
 class MemeTemplatesCollectionViewCell: UICollectionViewCell {
 
-    lazy var memeImageView: UIImageView = {
-        var imageView = UIImageView()
+    static let sizingCell = MemeTemplatesCollectionViewCell(frame: .zero)
```
Comment (mesbekmek) [inline Meme/Meme/MemeTemplatesCollectionViewCell.swift:13]: `static let` keyword in Swift means created once, but never mutated? A follow up to that, what's the difference between `static let` and `let`?

File: Meme/Meme/MemeTemplatesCollectionViewCell.swift (hunk 11 above)
Comment (vhart) [reply Meme/Meme/MemeTemplatesCollectionViewCell.swift:36]: Good question. It was just a way for me to ensure the layout

Comment (vhart) [reply Meme/Meme/MemeTemplatesCollectionViewCell.swift:36]: Tbh though this is missing a guard that the max isn’t the same max as the previous max, otherwise it’s a waste of an update

File: Meme/Meme/MemeTemplatesViewController.swift (hunk 12 above)
Comment (vhart) [reply Meme/Meme/MemeTemplatesViewController.swift:89]: Yep

File: Meme/Meme/MemeTemplatesViewController.swift (hunk 13 above)
Comment (vhart) [reply Meme/Meme/MemeTemplatesViewController.swift]: There’s a bug causing a crash in iOS 12 that necessitated this, but actually I wonder if that invalidate layout function you added to the flow layout removes the need for this now.

File: Meme/Meme/MemeTemplatesViewController.swift [hunk 16]
```diff
@@ -76,38 +75,74 @@ class MemeTemplatesViewController: UIViewController {
...
+
+        print("setneedslayout")
+        sizingCell.contentView.setNeedsLayout()
+        print("layoutifneeded")
+        sizingCell.contentView.layoutIfNeeded()
+
+        let size = sizingCell.contentView.systemLayoutSizeFitting(UIView.layoutFittingCompressedSize)
+        print("size: \(size)")
+
+        if sizingCell.memeImageView.didLoad {
```
Comment (mesbekmek) [inline Meme/Meme/MemeTemplatesViewController.swift:141]: Hmm this check seems a bit weird for me. Cause right now, we'll still return the size even if the this statement returns false, which indicates to me that in some cases, we will be returning the wrong size for the `sizeForItemAt indexPath` method. 

I understand this is more about saving the size calculation and returning it early at the beginning of this method if it exists already, but we still end up returning the wrong size in some cases right? i.e if the image is not loaded by the time we get to this check.

File: Meme/Meme/MemeTemplatesCollectionViewCell.swift (hunk 15 above)
Comment (vhart) [reply Meme/Meme/MemeTemplatesCollectionViewCell.swift:13]: let at the property level is an instance variable 
Static let is a namespace (class or struct or enum) level constant

File: Meme/Meme/MemeTemplatesViewController.swift [hunk 17]
```diff
@@ -76,38 +75,74 @@ class MemeTemplatesViewController: UIViewController {
...
 
     func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
-        return CGSize(width: collectionView.frame.width, height: 250)
-    }
+        if let size = sizing[indexPath] {
+            return size
+        }
+
+        let sizingCell = MemeTemplatesCollectionViewCell.sizingCell
+        sizingCell.prepareForReuse()
```
Comment (mesbekmek) [inline Meme/Meme/MemeTemplatesViewController.swift:125]: So by calling prepare for reuse we essentially nullify the cell image view's image and image URL properties right? And on line 130 below we set the imageURL for the memeImageView which will lead to us fetching the image from the cache or async again. 

We also do something similar in the `cellForItemAt` method. Do we need to be fetching here as well during cell size calculation?

File: Meme/Meme/MemeTemplatesViewController.swift (hunk 16 above)
Comment (vhart) [reply Meme/Meme/MemeTemplatesViewController.swift:141]: If the image hasn’t loaded the size is just the size of the placeholder image

File: Meme/Meme/MemeTemplatesViewController.swift [hunk 18]
```diff
@@ -76,38 +75,74 @@ class MemeTemplatesViewController: UIViewController {
...
+        sizingCell.maxWidth = itemWidth
+        if let imageURL = self.templates[indexPath.row].imageURL {
+            print("setting url: \(imageURL)")
+            sizingCell.memeImageView.imageURL = imageURL
+        }
+
+        print("setneedslayout")
+        sizingCell.contentView.setNeedsLayout()
+        print("layoutifneeded")
+        sizingCell.contentView.layoutIfNeeded()
```
Comment (mesbekmek) [inline Meme/Meme/MemeTemplatesViewController.swift:137]: So similar question to above, why do we need to call these methods together, and also here in particular?

File: Meme/Meme/MemeTemplatesViewController.swift (hunk 17 above)
Comment (vhart) [reply Meme/Meme/MemeTemplatesViewController.swift:125]: The nice thing about the async fetcher is that future calls won’t trigger off a fetch it’ll just queue up the callback

File: Meme/Meme/MemeTemplatesViewController.swift (hunk 18 above)
Comment (vhart) [reply Meme/Meme/MemeTemplatesViewController.swift:137]: We force the view to apply auto layout and then calculate the size below that

File: Meme/Meme/MemeTemplatesViewController.swift [hunk 19]
```diff
@@ -76,38 +75,74 @@ class MemeTemplatesViewController: UIViewController {
...
 
     func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        collectionView.layoutIfNeeded()
         let cell = self.collectionView.dequeueReusableCell(withReuseIdentifier: cellID, for: indexPath) as! MemeTemplatesCollectionViewCell
-        cell.memeImageView.image = nil
+
+        cell.maxWidth = itemWidth
+        cell.onAsyncLoad = { [weak self] in
+            self?.collectionViewLayout
+                .invalidateLayout()
```
Comment (mesbekmek) [inline Meme/Meme/MemeTemplatesViewController.swift:110]: I noticed we invalidate the collection view layout in quite a few places: When we get the templates, when the view will layout its subviews, and also here in `cellForItemAt`. Do we need to invalidate in all of them?

Comment (vhart) [reply Meme/Meme/MemeTemplatesViewController.swift:110]: We invalidate anytime the collection view needs to recalculate it’s cell sizes/context size/offset

**Comment by vhart** [comment]
There’s still some slowness 🙁😭

File: Meme/Meme/ImageCache.swift [hunk 20]
```diff
@@ -8,20 +8,210 @@
...
+ */
+
+import Foundation
+
+/// - Tag: AsyncFetcher
+class AsyncFetcher {
+    // MARK: Types
+
+    /// A serial `OperationQueue` to lock access to the `fetchQueue` and `completionHandlers` properties.
+    private let serialAccessQueue = OperationQueue()
```
Comment (mesbekmek) [inline Meme/Meme/ImageCache.swift:48]: What is the significance of adding a serialAccessQueue, aren't operation queues safe to use from multiple threads without creating additional locks to synchronize access ?

Later on when you add the completion handlers, is it important that they happen one at a time? Same for invoking the completion handlers when image is fetched, can't we invoke the operations in parallel?
And also for canceling, even if multiple cancelation requests come in for a given url, shouldn't it be fine since it's okay to access it from multiple threads?

File: Meme/Meme/MemeTemplatesViewController.swift [hunk 21]
```diff
@@ -54,15 +57,11 @@ class MemeTemplatesViewController: UIViewController {
...
-        //        self.tableView.delegate = self
-        //        self.tableView.dataSource = self
-        //        self.tableView.register(MemeTemplatesTableViewCell.self, forCellReuseIdentifier: cellID)
-        //        self.tableView.rowHeight = UITableView.automaticDimension
-        //        self.tableView.estimatedRowHeight = 250
 
         self.collectionView.delegate = self
         self.collectionView.dataSource = self
         self.collectionView.register(MemeTemplatesCollectionViewCell.self, forCellWithReuseIdentifier: cellID)
+        self.collectionViewLayout.estimatedItemSize = CGSize(width: itemWidth, height: 200)
```
Comment (mesbekmek) [inline Meme/Meme/MemeTemplatesViewController.swift:67]: I think you should also set the prefetchDataSource property on the collection view in order to be able to use `UICollectionViewDataSourcePrefetching ` protocol
i.e   `self.collectionView.prefetchDataSource = self`
The second bullet point here suggests doing that: https://developer.apple.com/documentation/uikit/uicollectionviewdatasourceprefetching

Comment (mesbekmek) [reply Meme/Meme/MemeTemplatesViewController.swift:67]: Although, I tried doing this locally after pulling your changes, and I didn't see any noticeable difference.

File: Meme/Meme/ImageCache.swift (hunk 9 above)
Comment (mesbekmek) [reply Meme/Meme/ImageCache.swift:107]: But should we be preventing other operations unrelated to this url from being started?

File: Meme/Meme/MemeTemplatesViewController.swift (hunk 17 above)
Comment (mesbekmek) [reply Meme/Meme/MemeTemplatesViewController.swift:125]: Ah fair point. 👍

**Comment by mesbekmek** [comment]
@vhart Yeah, there is some slowness initially when it's fetching for the first time, but it's buttery smooth while scrolling. That's nice at least. Try setting the prefetchDataSource property on the collection view on your end and see if that improves the slowness for you.

File: Meme/Meme/MemeTemplatesViewController.swift (hunk 21 above)
Comment (vhart) [reply Meme/Meme/MemeTemplatesViewController.swift:67]: Lol I forgot to set that to self. After all that set up.. haha 😓

**Comment by vhart** [comment]
@mesbekmek I applied some caching and removed unnecessary re-renders ie `collectionview.layoutIfNeeded`  Your invalidation function in the flow layout is all that was needed to prevent the crash I was seeing. This + caching the loading cell's size led to a dramatic performance increase! Works well - I think.

File: Meme/Meme/ImageCache.swift (hunk 9 above)
Comment (vhart) [reply Meme/Meme/ImageCache.swift:107]: It’s not a long running process so it should be fine. Full disclaimer though, this is the way they do it in their sample code

File: Meme/Meme/ImageCache.swift (hunk 20 above)
Comment (vhart) [reply Meme/Meme/ImageCache.swift:48]: The queues themselves are thread safe but writing to the callbacks dictionary is not. The serial queue will synchronize writes to the dictionary and then dispatch to the concurrent operation queue making sure the operations aren’t held up.

Comment (vhart) [reply Meme/Meme/ImageCache.swift:48]: Completion handlers can be invoked in parallel and imo should be dispatched to a global queue separate from the operation queues

**Review by mesbekmek — APPROVED** [review]
LGTM

**Comment by vhart** [comment]
@mesbekmek is the performance acceptable?

**Comment by mesbekmek** [comment]
Hey the performance is pretty good now 👍
